From 05f42d6a09b861a425541408ddac823dafdc5501 Mon Sep 17 00:00:00 2001
From: Pingfan Liu <piliu@redhat.com>
Date: Wed, 15 Feb 2023 19:23:40 +0800
Subject: [PATCH 749/921] [Backport] dm: add cond_resched() to dm_wq_work()

stable inclusion
from stable-v5.10.173
commit 3383f79d6b0a0641f767052fda3952dd1805d2d6
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I6PMQI

Reference: https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?id=3383f79d6b0a0641f767052fda3952dd1805d2d6

--------------------------------

commit 0ca44fcef241768fd25ee763b3d203b9852f269b upstream.

Otherwise the while() loop in dm_wq_work() can result in a "dead
loop" on systems that have preemption disabled. This is particularly
problematic on single cpu systems.

Cc: stable@vger.kernel.org
Signed-off-by: Pingfan Liu <piliu@redhat.com>
Acked-by: Ming Lei <ming.lei@redhat.com>
Signed-off-by: Mike Snitzer <snitzer@kernel.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Wang Hai <wanghai38@huawei.com>
---
 drivers/md/dm.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/md/dm.c b/drivers/md/dm.c
index e1240c174699..d2dd2708799c 100644
--- a/drivers/md/dm.c
+++ b/drivers/md/dm.c
@@ -2379,6 +2379,7 @@ static void dm_wq_work(struct work_struct *work)
 			break;
 
 		submit_bio_noacct(bio);
+		cond_resched();
 	}
 }
 
-- 
2.17.1

